[profiler] initialize kineto during lazy cuda initialization

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #97199

**Summary**: As suggested in @robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during lazy cuda init, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by recording the time before and after the kineto initialization call.
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

davidberard98 committed:
[WIP][profiler] initialize kineto before doing any profiling.

As suggested in @robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to torch initialization, so that it gets called before any cuda graphs are created.

Tested locally and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[WIP][profiler] initialize kineto before doing any profiling."


As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to torch initialization, so that it gets called before any cuda graphs are created.

Tested locally and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 commented:
cc @robieta for review as well, it's not letting me add you as a reviewer

aaronenyeshi reviewed (approved):
Looks great! Thanks for adding the repro to improve our tests!

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

robieta commented on torch/csrc/profiler/python/init.cpp:
```diff
@@ -293,6 +294,7 @@ void initPythonBindings(PyObject* module) {
   m.def(
       "_disable_execution_graph_observer",
       &torch::profiler::impl::disableExecutionGraphObserver);
+  m.def("_initialize_kineto", &torch::profiler::impl::kineto::initializeKineto);
```
Why not just do this in C++ during runtime init? (I just did it through Python because it was expedient.)

davidberard98 replied:
thanks for the suggestion - also, I think I will try to move this into lazy cuda initialization (the ROCM & windows tests are failing because this is somehow triggering cuda init during `import torch`)

robieta reviewed (approved):
A couple minor nits, but overall looks good!

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

robieta commented on test/profiler/test_profiler.py:
```diff
@@ -138,6 +138,39 @@ def custom_layer(input_ten):
+        try:
+            # repro taken from #75504.
+            # Launch in a separate process to catch hanging/illegal memory errors
+            p = subprocess.Popen([sys.executable, "-c", """
```
Mostly fbcode. I've seen cases in OSS where this pattern doesn't work, but IIRC they were fairly niche. (And I don't recall a specific example)

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 committed:
remove pyi changes on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 committed:
check isProfilerRegistered() in test on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during import"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during import, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
try reordering things to fix windows build error on "[profiler] initialize kineto during lazy cuda initialization"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during lazy cuda init, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
fix typo on "[profiler] initialize kineto during lazy cuda initialization"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during lazy cuda init, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by adding `time()` calls before and after the call to `_profiler_init` in `torch/profiler/__init__.py`
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 committed:
Update on "[profiler] initialize kineto during lazy cuda initialization"


**Summary**: As suggested in robieta in #75504, this workaround seems to fix the cudagraph profiling issues. It moves the kineto initialization step to happen during lazy cuda init, so that kineto initialization gets called before any cuda graphs are created.

**Tests**: 
* Tested locally (in OSS environment) and verified that the issue goes away (although - locally, the symptom is a hanging process, not an illegal memory access).
* Added a test `test_cudagraph_profiling` - which runs in a separate process, so that it won't hang the test suite if the hanging behavior continues.
* Latency: locally, this adds 51ms to the import time. This is tested by recording the time before and after the kineto initialization call.
* fbcode sanity check to make sure this still works internally: D44528790

Differential Revision: [D44241459](https://our.internmc.facebook.com/intern/diff/D44241459)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44241459).

davidberard98 commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -1183,6 +1184,9 @@ static PyObject* THCPModule_initExtension(PyObject* self, PyObject* noargs) {
   set_module_attr("default_generators", default_cuda_generators);
   bindGetDeviceProperties(m);
 
+  // initialize kineto to fix #75504
+  torch::profiler::impl::kineto::initializeKineto();
```
cc @ngimel - is this the right place to put this in order to initialize kineto during lazy cuda initialization?

ngimel replied:
I'm not entirely sure - it might initialize context on the GPU before it needs to be initialized, which we don't want to happen.
